Libraries/Opc.Ua.Configuration/ApplicationConfigurationBuilder.cs

```diff
@@ -0,0 +1,880 @@
+using System.Xml;
+
+namespace Opc.Ua.Configuration
+{
+    /// <summary>
+    /// A class that builds a configuration for a UA application.
+    /// </summary>
+    public class ApplicationConfigurationBuilder :
+        IApplicationConfigurationBuilder
+    {
+        #region ctor
+        /// <summary>
+        /// Create the application instance builder.
+        /// </summary>
+        public ApplicationConfigurationBuilder(ApplicationInstance applicationInstance)
+        {
+            ApplicationInstance = applicationInstance;
+        }
+        #endregion
+
+        #region Public Properties
+        /// <summary>
+        /// The application instance used to build the configuration.
+        /// </summary>
+        public ApplicationInstance ApplicationInstance { get; private set; }
+        /// <summary>
+        /// The application configuration.
+        /// </summary>
+        public ApplicationConfiguration ApplicationConfiguration => ApplicationInstance.ApplicationConfiguration;
+        #endregion
+
+        #region Public Methods
+        /// <inheritdoc/>
+        public IApplicationConfigurationBuilderClientSelected AsClient()
+        {
+            switch (ApplicationInstance.ApplicationType)
+            {
+                case ApplicationType.Client:
+                case ApplicationType.ClientAndServer:
+                    break;
+                case ApplicationType.Server:
+                    ApplicationInstance.ApplicationType =
+                        m_typeSelected ? ApplicationType.ClientAndServer : ApplicationType.Client;
+                    break;
+                default:
+                    throw new ArgumentException("Invalid application type for client.");
+            }
+
+            m_typeSelected = true;
+
+            ApplicationConfiguration.ClientConfiguration = new ClientConfiguration();
+
+            return this;
+        }
+
+        /// <inheritdoc/>
+        public IApplicationConfigurationBuilderSecurityOptions AddSecurityConfiguration(
+            string subjectName,
+            string pkiRoot = null,
+            string appRoot = null,
+            string rejectedRoot = null
+            )
+        {
+            pkiRoot = DefaultPKIRoot(pkiRoot);
+            appRoot = DefaultPKIRoot(appRoot);
+            rejectedRoot = DefaultPKIRoot(rejectedRoot);
+            var appStoreType = CertificateStoreIdentifier.DetermineStoreType(appRoot);
+            var pkiRootType = CertificateStoreIdentifier.DetermineStoreType(pkiRoot);
+            var rejectedRootType = CertificateStoreIdentifier.DetermineStoreType(rejectedRoot);
+            ApplicationConfiguration.SecurityConfiguration = new SecurityConfiguration
+            {
+                // app cert store
+                ApplicationCertificate = new CertificateIdentifier()
+                {
+                    StoreType = appStoreType,
+                    StorePath = DefaultCertificateStorePath(TrustlistType.Application, pkiRoot),
```

Comment:
should be _**appRoot**_ here instead of pkiRoot, right?
`StorePath = DefaultCertificateStorePath(TrustlistType.Application, appRoot),`

Reply:
Indeed, good catch! Also if appRoot == null, the pkiRoot should be used